### PR TITLE
ethdb, core: implement delete for db batch

### PR DESCRIPTION
**Pull request by rjl493456442**

### CODE REVIEW

File: ethdb/memory_database.go [hunk 1]
```diff
@@ -110,11 +110,23 @@ func (b *memBatch) Put(key, value []byte) error {
 	return nil
 }
 
+func (b *memBatch) Delete(key []byte) error {
+	b.db.lock.Lock()
```
Comment (holiman) [inline ethdb/memory_database.go]: How come you use the `b.db.lock.Lock()` here, when it's not used for `Put` ?

File: light/txpool.go [hunk 2]
```diff
@@ -504,15 +506,17 @@ func (self *TxPool) Content() (map[common.Address]types.Transactions, map[common
...
 	for _, tx := range txs {
-		//self.RemoveTx(tx.Hash())
 		hash := tx.Hash()
 		delete(self.pending, hash)
-		self.chainDb.Delete(hash[:])
+		batch.Delete(hash.Bytes())
 		hashes = append(hashes, hash)
 	}
 	self.relay.Discard(hashes)
+	batch.Write()
```
Comment (karalabe) [inline light/txpool.go]: I'm not really sure what `self.relay.Discard(hashes)`, but I think it's better to write the changes first and then call it. That way we retain the behavior that the writes conclude before calling this method.

File: core/headerchain.go [hunk 3]
```diff
@@ -448,22 +450,23 @@ func (hc *HeaderChain) SetHead(head uint64, delFn DeleteCallback) {
...
+		rawdb.DeleteTd(batch, hash, num)
 
 		hc.currentHeader.Store(hc.GetHeader(hdr.ParentHash, hdr.Number.Uint64()-1))
 	}
 	// Roll back the canonical chain numbering
 	for i := height; i > head; i-- {
-		rawdb.DeleteCanonicalHash(hc.chainDb, i)
+		rawdb.DeleteCanonicalHash(batch, i)
 	}
+	batch.Write()
```
Comment (karalabe) [inline core/headerchain.go]: Please add a newline here after this.

File: core/headerchain.go [hunk 4]
```diff
@@ -156,13 +156,15 @@ func (hc *HeaderChain) WriteHeader(header *types.Header) (status WriteStatus, er
...
+		batch := hc.chainDb.NewBatch()
 		for i := number + 1; ; i++ {
 			hash := rawdb.ReadCanonicalHash(hc.chainDb, i)
 			if hash == (common.Hash{}) {
 				break
 			}
-			rawdb.DeleteCanonicalHash(hc.chainDb, i)
+			rawdb.DeleteCanonicalHash(batch, i)
 		}
+		batch.Write()
```
Comment (karalabe) [inline core/headerchain.go]: Please add a newline after this,

File: core/blockchain.go [hunk 5]
```diff
@@ -1340,9 +1340,11 @@ func (bc *BlockChain) reorg(oldBlock, newBlock *types.Block) error {
 	diff := types.TxDifference(deletedTxs, addedTxs)
 	// When transactions get deleted from the database that means the
 	// receipts that were created in the fork must also be deleted
+	batch := bc.db.NewBatch()
 	for _, tx := range diff {
-		rawdb.DeleteTxLookupEntry(bc.db, tx.Hash())
+		rawdb.DeleteTxLookupEntry(batch, tx.Hash())
 	}
+	batch.Write()
```
Comment (karalabe) [inline core/blockchain.go]: Please add a newline after this.

**Comment by rjl493456442** [comment]
@karalabe Updated.

**Pull request merged**